[CORL-409] Prevent users from ignoring staff members

nick-funk opened the pull request:
## What does this PR do?

This removes the ignore button on staff member's overlay popup.
This also validates ignore requests in the backing API to prevent users from ignoring staff members.

## How do I test this PR?

Front-End:
- Create a user with a staff role (Admin, Mod, or Staff).
- Comment on something in any test discussion.
- Create a regular user that is not a staff role.
- View the previous discussion comment, click on the staff user's name.
- The ignore button should not be visible on the staff user's overlay popup/card.

Back-end:
- Request an ignore for a user that is a staff member
- Should return a `USER_CANNOT_BE_IGNORED`

nick-funk committed:
Throw error if user tries to ignore a staff member

Throws a UserCannotBeIgnoredError if a user tries to ignore
a user who is a staff member. A staff member in this case is
considered anyone who has a role of staff, moderator, or admin.

CORL-409

nick-funk committed:
Prevent users from ignoring staff in the user info popover

Creates the staff roles in a constant next to the user model.
Uses this to add a computed property to the user resolver.

CORL-409

nick-funk commented:
Looking into the tests and adding some more now. Tests will likely fail, just pushing up for any preliminary feedback/transparency.

wyattjoh commented on src/core/server/models/user/helpers.ts:
```diff
@@ -0,0 +1,11 @@
+
+import { User } from ".";
+
+export async function userIsStaff(user: User) {
```
Function does not use `await`, therefore you can remove the `async` term. Also, scoping is helpful here:

```suggestion
export function userIsStaff(user: Pick<User, "role">) {
```

nick-funk replied:
Fixed this up, removed the async and also the await.

wyattjoh commented on src/core/server/services/users/index.ts:
```diff
@@ -585,6 +587,11 @@ export async function ignore(
     throw new UserNotFoundError(userID);
   }
 
+  const userToBeIgnoredIsStaff = await userIsStaff(targetUser);
```
Assuming we can remove `async`, remove the `await`, and probably just put it into the `if (...) {` clause as the value isn't used outside this check.

nick-funk replied:
You're right, I've removed the async and also the await.

nick-funk committed:
Remove unnecessary async declaration from userIsStaff helper function

CORL-409

nick-funk committed:
Specify ignoreable on users in client test fixtures

Allows the tests to pass for the required computed property
of ignoreable that is computed by whether a user is a staff
member or not.

CORL-409

nick-funk committed:
Update more fixtures with ignoreable property on mocked users/commenters

CORL-409

nick-funk committed:
Consolidate ignore-able calculation into re-usable helper methods

Re-use the logic for whether a role is a staff member to
clearly define when a user is ignore-able or not across the
business logic.

CORL-409

nick-funk committed:
Set the ignoreable optimisticResponse on comment mutations

We have set the ignoreable value in the graphQL schema, so now
the optimisticResponses are looking for a default value to use
until the data result arrives. Put to false since the ignoreable
value is set on our author, we likely don't want to ignore ourselves.

CORL-409
